### PR TITLE
CAMEL-12295: Added an option to Jms Endpoints so that they format JMS…

### PR DESCRIPTION
**Pull request by aldettinger**
… date properties according to the ISO 8601 standard.

### CODE REVIEW

**Comment by aldettinger** [comment]
Well, at least we need one higher level test to check that the option is well parsed.

File: components/camel-jms/src/test/java/org/apache/camel/component/jms/JmsBindingTest.java [hunk 1]
```diff
@@ -18,32 +18,69 @@
...
+        Object value = jmsBindingUnderTest.getValidJMSHeaderValue("foo", Date.from(instant));
+        assertNotNull(value);
+        // We can't assert further as the returned value is bound to the machine time zone and locale
+    }
+
+    @Test
+    public void testGetValidJmsHeaderValueWithIso8601DateShouldSucceed() {
+        when(mockJmsConfiguration.isFormatDatePropertiesToIso8601()).thenReturn(true);
+        Object value = jmsBindingUnderTest.getValidJMSHeaderValue("foo", Date.from(instant));
+        assertEquals("2018-02-26T19:12:18Z", value);
```
Comment (aldettinger) [inline components/camel-jms/src/test/java/org/apache/camel/component/jms/JmsBindingTest.java]: Expressing date at GMT+0 is not necessarily easy to read for end users deploying in a single country. But, I find them easier to work with when your deployments is spread across data centers located in different time zones.

File: components/camel-jms/src/main/java/org/apache/camel/component/jms/JmsConfiguration.java [hunk 2]
```diff
@@ -491,6 +491,9 @@
         + " By enabling this option the message payload is read into memory in chunks and each chunk is then written to the StreamMessage until no more data.")
     private boolean streamMessageTypeEnabled;
 
+    @UriParam(label = "producer", description = "Sets whether JMS date properties should be formatted according to the ISO 8601 standard.")
+    private boolean formatDatePropertiesToIso8601;
```
Comment (davsclaus) [inline components/camel-jms/src/main/java/org/apache/camel/component/jms/JmsConfiguration.java]: We should call it headers as that is what they are stored as in Camel, eg formatDateHeadersToIso...
And you should have this option on the JmsComponent also so you can configure it once only (eg globally).

Comment (aldettinger) [reply components/camel-jms/src/main/java/org/apache/camel/component/jms/JmsConfiguration.java]: Thanks for input Claus. I will do so.

**Comment by aldettinger** [comment]
Ok, so I've added the option at component level also, changed the name and completed with a route level test. Should I merge it in 2.21 ?

**Comment by davsclaus** [comment]
Yeah but you should also add the `@ManagedAttrbiute` to the `JmsEndpoint` class like the other options have.

**Comment by davsclaus** [comment]
PS: You may also want to rebuild camel-amqp as it extends camel-jms and its docs/spring-boot-starter needs to be updated as well, but this happens anyway on new rebuild of all the code - but just wanted to let you know

**Comment by aldettinger** [comment]
The new option is now exposed via JMX and camel-ampq is building as well. Let's take time to review this a bit.

**Comment by oscerd** [comment]
Looks good to me, but lets wait for the final ok from @davsclaus

**Comment by aldettinger** [comment]
Ok. Thanks for the review guys.

**Pull request closed without merging**